### PR TITLE
ARROW-1638: [Java] IPC roundtrip for null type

### PR DESCRIPTION
**Pull request by tianchen92**
Related to [ARROW-1638](https://issues.apache.org/jira/browse/ARROW-1638).

### CODE REVIEW

File: java/vector/src/main/java/org/apache/arrow/vector/NullVector.java
Comment (emkornfield) [inline java/vector/src/main/java/org/apache/arrow/vector/NullVector.java]: this should ensure the size is reset?

File: java/vector/src/main/java/org/apache/arrow/vector/NullVector.java
Comment (emkornfield) [inline java/vector/src/main/java/org/apache/arrow/vector/NullVector.java]: should these be defined for other null vectors?

File: java/vector/src/main/java/org/apache/arrow/vector/VectorLoader.java
Comment (emkornfield) [inline java/vector/src/main/java/org/apache/arrow/vector/VectorLoader.java]: remove this?

**Review by emkornfield — CHANGES_REQUESTED** [review]
I think you might need to update JsonFileWriter, JsonFileReader, etc.

File: java/vector/src/main/java/org/apache/arrow/vector/NullVector.java
Comment (tianchen92) [reply java/vector/src/main/java/org/apache/arrow/vector/NullVector.java]: IMO, what does this mean?

**Comment by emkornfield** [comment]
Sorry, In my opinion.

**Comment by tianchen92** [comment]
> Sorry, In my opinion.

Something is wrong?

**Comment by emkornfield** [comment]
Nothing wrong, I was apologizing for using the abbreviation.

**Comment by tianchen92** [comment]
> Nothing wrong, I was apologizing for using the abbreviation.

Sorry for misleading, I just want to ask what do your mean by "should these be defined for other null vectors?" above, forgive me for poor English :)

**Comment by tianchen92** [comment]
PR updated, I am not quite familiar with this part, please let me know if you have other comments, thanks

File: java/vector/src/main/java/org/apache/arrow/vector/NullVector.java
Comment (tianchen92) [reply java/vector/src/main/java/org/apache/arrow/vector/NullVector.java]: seems here only copies value and not set valueCount. transfer and splitAndTransfer changes valueCount.

File: java/vector/src/main/java/org/apache/arrow/vector/ZeroVector.java
Comment (emkornfield) [inline java/vector/src/main/java/org/apache/arrow/vector/ZeroVector.java]: this seems like a bug in the original implementation, it seems that this should probably thow an index out of bounds exception?

**Comment by emkornfield** [comment]
@tianchen92 the real test for this would be adding it to the python integration test (integration/integration_test.py)  if you don't feel comfortable doing so i can make the change and push to your branch in the next few days.

**Comment by tianchen92** [comment]
> @tianchen92 the real test for this would be adding it to the python integration test (integration/integration_test.py) if you don't feel comfortable doing so i can make the change and push to your branch in the next few days.

Really appreciate for your help, I have invited you as collaborator(https://github.com/tianchen92/arrow/invitations), thanks!

**Comment by pitrou** [comment]
Right. There should simply be a clear and unambiguous specification.

**Comment by tianchen92** [comment]
@emkornfield I guess you probably forgot to add python integration test?

**Comment by emkornfield** [comment]
> @emkornfield I guess you probably forgot to add python integration test? 
Still on my todo list.

**Comment by emkornfield** [comment]
It appears this needs to be rebased?

**Comment by tianchen92** [comment]
> It appears this needs to be rebased?

Sure, rebased.

**Comment by wesm** [comment]
Now that I've just merged ARROW-6379 you may want to update this to correspond to that

**Comment by emkornfield** [comment]
I should be able to get to an integration test next week.

**Comment by wesm** [comment]
What is the status of this patch? Needs to be rebased.

**Comment by tianchen92** [comment]
I just rebased, if @emkornfield has no time recently, could we merge this first since this pending a long time and seems there's already a JIRA for integration test(ARROW-1636)

**Comment by emkornfield** [comment]
Yes, we can handle integration tests separately.  I'll do another quick pass today and then merge.

File: java/adapter/avro/src/main/java/org/apache/arrow/consumers/AvroNullConsumer.java
Comment (emkornfield) [inline java/adapter/avro/src/main/java/org/apache/arrow/consumers/AvroNullConsumer.java]: should there be an increment on NullVector count when consuming?

File: java/vector/src/main/java/org/apache/arrow/vector/NullVector.java
Comment (emkornfield) [inline java/vector/src/main/java/org/apache/arrow/vector/NullVector.java]: this should probably be IntMax?

File: java/vector/src/main/java/org/apache/arrow/vector/NullVector.java
Comment (emkornfield) [inline java/vector/src/main/java/org/apache/arrow/vector/NullVector.java]: chould be more succinct with checkArgument.

File: java/vector/src/main/java/org/apache/arrow/vector/NullVector.java
Comment (emkornfield) [inline java/vector/src/main/java/org/apache/arrow/vector/NullVector.java]: it probably makes sense to have this be some none-zero value, so it contributes to changing the hash-value in degenerate cases

File: java/vector/src/main/java/org/apache/arrow/vector/NullVector.java
Comment (emkornfield) [inline java/vector/src/main/java/org/apache/arrow/vector/NullVector.java]: I think this needs to set value count on the target if toIndex is larger then the current index?

File: java/vector/src/main/java/org/apache/arrow/vector/ZeroVector.java
Comment (emkornfield) [inline java/vector/src/main/java/org/apache/arrow/vector/ZeroVector.java]: if valueCount > 0 should it be an error?

File: java/vector/src/main/java/org/apache/arrow/vector/ZeroVector.java
Comment (emkornfield) [inline java/vector/src/main/java/org/apache/arrow/vector/ZeroVector.java]: should be an error if from/to are greater then zero?

File: java/vector/src/test/java/org/apache/arrow/vector/ipc/TestArrowReaderWriter.java
Comment (emkornfield) [inline java/vector/src/test/java/org/apache/arrow/vector/ipc/TestArrowReaderWriter.java]: is the intVector necessary here?

File: java/vector/src/test/java/org/apache/arrow/vector/ipc/BaseFileTest.java
Comment (emkornfield) [inline java/vector/src/test/java/org/apache/arrow/vector/ipc/BaseFileTest.java]: might be good to vary valueCount.

**Review by emkornfield — CHANGES_REQUESTED** [review]
still some changes on the java side, I think once these are resolve/answered we can merge and save integration test for later.

File: java/vector/src/main/java/org/apache/arrow/vector/ZeroVector.java
Comment (tianchen92) [reply java/vector/src/main/java/org/apache/arrow/vector/ZeroVector.java]: I prefer not add check here since it dose nothing here and more important this will cause some unexpected exception which works fine before (i.e. if we create a ListVector and setValueCount it will fail since it's initialized with ZeroVector)

File: java/vector/src/main/java/org/apache/arrow/vector/ZeroVector.java
Comment (tianchen92) [reply java/vector/src/main/java/org/apache/arrow/vector/ZeroVector.java]: Since ZeroVector has no values, any index is invalid include 0, so just do nothing in this method.

File: java/adapter/avro/src/main/java/org/apache/arrow/consumers/AvroNullConsumer.java
Comment (tianchen92) [reply java/adapter/avro/src/main/java/org/apache/arrow/consumers/AvroNullConsumer.java]: Right, fixed.

File: java/vector/src/main/java/org/apache/arrow/vector/NullVector.java
Comment (tianchen92) [reply java/vector/src/main/java/org/apache/arrow/vector/NullVector.java]: Here should be valueCount I think according to other vectors implementation, fixed.

File: java/vector/src/main/java/org/apache/arrow/vector/NullVector.java
Comment (tianchen92) [reply java/vector/src/main/java/org/apache/arrow/vector/NullVector.java]: fixed.

File: java/vector/src/main/java/org/apache/arrow/vector/NullVector.java
Comment (tianchen92) [reply java/vector/src/main/java/org/apache/arrow/vector/NullVector.java]: fixed.

File: java/vector/src/main/java/org/apache/arrow/vector/NullVector.java
Comment (tianchen92) [reply java/vector/src/main/java/org/apache/arrow/vector/NullVector.java]: fixed.

File: java/vector/src/test/java/org/apache/arrow/vector/ipc/TestArrowReaderWriter.java
Comment (tianchen92) [reply java/vector/src/test/java/org/apache/arrow/vector/ipc/TestArrowReaderWriter.java]: removed.

File: java/vector/src/test/java/org/apache/arrow/vector/ipc/BaseFileTest.java
Comment (tianchen92) [reply java/vector/src/test/java/org/apache/arrow/vector/ipc/BaseFileTest.java]: fixed.

File: java/vector/src/main/java/org/apache/arrow/vector/ZeroVector.java
Comment (emkornfield) [reply java/vector/src/main/java/org/apache/arrow/vector/ZeroVector.java]: I guess by that argument an exception should be thrown no matter what?

Comment (emkornfield) [reply java/vector/src/main/java/org/apache/arrow/vector/ZeroVector.java]: but this seems to be existing behavior.

**Comment by emkornfield** [comment]
+1 thank you.

**Pull request closed without merging**
